### ISSUE TITLE
Binary exporting and importing values

### ISSUE THREAD

**Issue by sattel** [bluescarni/mppp#109]
Hi!

I want to store/restore integers and rationals to/from a buffer. For now I use mpz_export() and mpz_import(), but this is not the most efficient way for small values as it constructs temporary objects. Using the string representation is also not very efficient. So what way do you recommend for storing/restoring the numbers?

Did I miss something? Or is it just a feature request :-)

In general I have to say this is the just the best C++ library for large integers and rationals...

Thanks for it,
sattel

**Comment by bluescarni** [comment]
Hello @sattel,

you are not missing anything, serialization support is indeed not implemented yet, but it is on the roadmap.

I have been thinking about it for a while, but my thoughts so far have been focused on enabling (de)serialization of mp++ objects in popular serialization libraries such as Boost.Serialization, cereal and msgpack-c (of course, this type of serialization support would be optional in order to avoid too many mandatory dependencies).

However, you bring up an interesting point about having builtin fast binary serialization not relying on any 3rd party code. So maybe we can discuss a bit about it.

What type of interface would you prefer? Do you need the full breadth of options provided by GMP's ``mpz_import/export()``? I have no experience with these functions, so it is not clear to me how much the support for, e.g., different types of endianness is useful in practice. My instinct would be to provide (at least initially) a fast non-portable binary serialization support.

Would you prefer to be able to (de)serialize directly to a memory buffer (i.e., a ``unsigned char *`` I presume), or would you prefer having an interface based on C++ streams? I feel like probably both interfaces should be available to cater to most use cases.

Let me know if you have any thoughts about this.

And thanks for the kind words, they are much appreciated.

**Comment by sattel** [comment]
Hi Francesco!

Thanks for the quick and kind reply.

I also don't think we need all features of mpz_import()/mpz_export(). I just use it for serializing the data into a simple bytestream. By the way: the sign bit is not handled by these functions, just the absolute value, but that's OK.

Currenty I do this for serializing:

```cpp
mppp::integer<1> v = some_big_integer;

// --- serialize big integer into buffer vector
const auto& mpz_view = v.get_mpz_view();
auto bufsize = (mpz_sizeinbase(mpz_view, 2) + 7) / 8;
std::vector<uint8_t> buffer(bufsize);
(void) mpz_export((void*) buffer.data(), &bufsize, 1, 1, 0, 0, mpz_view);
```

... and this for deserializing:

```cpp
std::vector<uint8_t> buffer; // container serialized big integer

// --- deserialize big integer from buffer vector
mpz_t _v;
mpz_init(_v);
(void) mpz_import(_v, buffer->size(), 1, 1, 0, 0, buffer->data());
mppp::integer<1> v(std::move(_v));
```

I would prefer a very simple interface like this for serializing:

```cpp
mppp::integer<1> v = some_big_integer;

// --- serialize big integer into buffer vector
auto bufsize = (v.nbits() + 7) / 8;
std::vector<uint8_t> buffer(bufsize);
(void) v.serialize(buffer->data(), buffer->size());
```

... and this for deserializing:

```cpp
std::vector<uint8_t> buffer; // container serialized big integer

// --- deserialize big integer from buffer vector
mppp::integer<1> v = mppp::integer<1>::deserialize(buffer->data(), buffer->size());
```

Another option would be to put der buffer pointer and size into the constructor of the big integer...

The return value could be the number of written bytes or an errorcode if the buffer was too small.

So these are my ideas for an initial interface. This would provide the feature itself. I'm not quite sure about the handling of the sign bit (Putting it in the bytestream or let the user manage this information...)

What do you think?

**Comment by bluescarni** [comment]
I think it's reasonable and very much similar to what I had in mind. I'd like to have perhaps multiple overloads for the serialization functions, in order to allow using both an ``std::vector`` (or maybe even an ``std::array``) and raw pointers (these allow, for instance, to write directly to a file via memory mapping).

I think we need also some utility functions that compute the number of bytes needed for serialization (so that one can pre-allocate a buffer beforehand, as you show in the examples). I also think that it's just better to store the sign as well, I don't think it's gonna be a problem performance- or storage-wise.

Give me a few days to think about it a bit more and I'll ping you back when I have a pull request ready.

**Comment by sattel** [comment]
This sounds great. I like your ideas and I'm really looking forward to test those additions.

**Comment by bluescarni** [comment]
@sattel There's an initial implementation up at #110. Currently it does not cover stream-based serialisation, only buffer-based.

I still need to write the tutorial, but API docs and tests are there. Let me know what you think if you want to take a look at it.

**Comment by bluescarni** [comment]
@sattel I have added stream-based serialisation as well. Only the tutorial bits are missing now, once they are written I'll consider this ready for merging.

**Comment by bluescarni** [comment]
I have merged #110, I am going ahead and close this.
